simulariumio/cellpack/cellpack_converter.py

```diff
@@ -0,0 +1,350 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+
+import logging
+import os
+import numpy as np
+import json
+from scipy.spatial.transform import Rotation as R
+
+from ..constants import DISPLAY_TYPE
+from ..data_objects.camera_data import CameraData
+from ..trajectory_converter import TrajectoryConverter
+from ..data_objects import TrajectoryData, AgentData, DimensionData
+from ..data_objects import MetaData, DisplayData
+from .cellpack_data import HAND_TYPE, CellpackData
+from cellpack.autopack.iotools_simple import RecipeLoader
+
+###############################################################################
+
+log = logging.getLogger(__name__)
+
+###############################################################################
+
+
+class CellpackConverter(TrajectoryConverter):
+    def __init__(self, input_data: CellpackData):
+        """
+        This object reads packing results outputs
+        from Cellpack (http://www.cellpack.org)
+        and plot data and writes them in the JSON format used
+        by the Simularium viewer
+
+        Parameters
+        ----------
+        input_data : CellpackData
+            An object containing info for reading
+            Cellpack simulation trajectory outputs and plot data
+        """
+        self._data = self._read(input_data)
+
+    @staticmethod
+    def _get_box_center(recipe_data):
+        options = recipe_data["options"]
+        bb = options["boundingBox"]
+        x_pos = (bb[1][0] - bb[0][0]) / 2.0 + bb[0][0]
+        y_pos = (bb[1][1] - bb[0][1]) / 2.0 + bb[0][1]
+        z_pos = (bb[1][2] - bb[0][2]) / 2.0 + bb[0][2]
+        return [
+            x_pos,
+            y_pos,
+            z_pos,
+        ]
+
+    @staticmethod
+    def _get_boxsize(recipe_data):
+        options = recipe_data["options"]
+        bb = options["boundingBox"]
+        x_size = bb[1][0] - bb[0][0]
+        y_size = bb[1][1] - bb[0][1]
+        z_size = bb[1][2] - bb[0][2]
+        return [
+            x_size,
+            y_size,
+            z_size,
+        ]
+
+    @staticmethod
+    def _get_euler_from_matrix(data_in, handedness):
+        rotation_matrix = [data_in[0][0:3], data_in[1][0:3], data_in[2][0:3]]
+        if handedness == HAND_TYPE.LEFT:
+            euler = R.from_matrix(rotation_matrix).as_euler("ZYX", degrees=False)
```

Comment:
did you consider transforming the matrix (or quaternion) itself and leaving all the from_matrix and as_euler to be the standard way?

Reply:
I also found this other solution which at first glance looks different than yours:  https://stackoverflow.com/questions/31191752/right-handed-euler-angles-xyz-to-left-handed-euler-angles-xyz

Reply:
but then again, as i recall, you said this was all working so why mess with it?  But it would be good to understand foundationally why it is correct, for future debugging etc